File: PhraseMaker.py
import random, math, string
from operator import itemgetter

# Size of the population.
# Not related to the length of the target string.
popSize = 256

# Chance out of 1 that each letter will be randomised.
mutability = 0.02

genNum = 0 # The number of generations iterated.

finalPhrase = "To be or not to be, that is the question." # Target phrase.
# Moved to bottom to be taken as input, only use this phrase if (no input is happen?) <- How are no input is happen(?)
phraseLen = len(finalPhrase) # Set the length of the string.

alphabetu = string.ascii_uppercase
alphabetl = string.ascii_lowercase
punct = string.punctuation
digits = string.digits
chooseFrom = alphabetu + ' ' + alphabetl + punct + digits# String to choose randomly from.

# Generate a random population of n phrases, each of length phraseLen.
def populate(length):
    temp = []
    for i in range(length):
        creature = ''.join(random.sample(chooseFrom, phraseLen))
        temp.append(creature)
    genNum = 1
    return temp

# Toggle the case of a letter passed.
def toggleCase(l):
    if l.upper() == l:
        return l.lower()
    return l.upper()
# This function is never used, but I could use it to make the correct letter
# in the wrong case worth some fitness, if theis function was enabled for mutation.

# Function to calculate the value of a given phrase and the target phrase.
def evaluate(phrase, goal):
    fitness = 0
    for i in range(len(phrase)):
        if phrase[i] == goal[i]:
            fitness += 1
    fitness = fitness**2
    return fitness

# Function to reproduce given two creatures.
# Uses simple crossover of first half of first creature and second half of second creature.
def reproduce(phrase1, phrase2):
    wordLen = len(phrase1)
    if wordLen != len(phrase2):
        return phrase1
    offspring = str()
    for i in range(wordLen):
        if i < math.ceil(wordLen/2):
            offspring += phrase1[i]
        else:
            offspring += phrase2[i]
    offspring = mutate(offspring, mutability)
    return offspring

# Mutate a word using a passed mutability multiplier.
def mutate(word, chance):
    wordList = list(word)
    mutateList = []
    if chance != 0:
        for i in range(int(math.floor(1/chance))-1):
            mutateList.append(False)
    mutateList.append(True)
    for i in range(len(word)):
        mutateBool = random.choice(mutateList)
        if mutateBool:
            wordList[i] = random.choice(chooseFrom)
    return ''.join(wordList)

# Kill the last half of a population of creatures.
def killWorst(pop):
    killNum = int(math.floor(len(pop)/2))
    survivors = pop
    for i in range(killNum):
        del(survivors[len(pop) - killNum])
    return survivors

# Get a random pair for 'mating'.
def getPair(valpop):
    matingPool = []
    for i in valpop:
        for chance in range(i[0]):
            matingPool.append(i[1])
    pair = random.sample(matingPool, 2)
    return pair

# Load and return the next generation of 'creatures'.
def nextGen(pop):
    valuedPop = [] # TODO: make the previous gen's valuedPop persist so we only have to evaluate half.
    population = pop
    for word in population:
        fitness = evaluate(word, finalPhrase) # Evaluate the fitness of the word.
        valuedPop.append((fitness, word)) # Add the fitness and word to the list.
    valuedPop.sort(key=itemgetter(0), reverse=True) # Sort best to worst.
    print("Best: " + str(valuedPop[0][1]))
    print("Best fitness: " + str(valuedPop[0][0]))
    print("Generation: " + str(genNum + 1))
    del(population)
    population = []
    for i in valuedPop:
        population.append(i[1])
    # If the program has finished.
    if valuedPop[0][1] == finalPhrase:
        return population, True
    population = killWorst(population) # Kill off the weaklings.
    valuedPop = killWorst(valuedPop)
    population2 = []
    # Repopulate the recently 'freed' slots in the population.
    for i in range(int(math.floor(len(population)/2))):
        # Generate an offspring per 2 parents.
        pair = getPair(valuedPop)
        population2.append(reproduce(pair[0], pair[1]))
        pair = getPair(valuedPop)
        # Do it twice to generate 2 offspring per 2 parents.
        # This returns us to the original population size.
        population2.append(reproduce(pair[0], pair[1]))
    for i in population2:
        population.append(i)
    # Mutate everything again.
    for i in range(len(population)):
        population[i] = mutate(population[i], mutability)
    return population, False


# Do stuff here.
finalPhrase = str(raw_input("Phrase to search for: ")) # Target phrase from user input.
phraseLen = len(finalPhrase) # Set the length of the target.
population = populate(popSize) # Make the first generation.
fin = False
while not fin:
    population, fin = nextGen(population) # Iterate to the next generation
    genNum += 1
print("Completed, iterated over " + str(genNum) + " generations.")

File: PhraseMaker.js
var popSize = 512; // How many creatures in the population.
var mutability = 0.02; // Probability that any letter of 'DNA' will mutate at a time.
var genNum = 0; // Used to keep track of the generation number.
var finalPhrase = "To be or not to be, that is the question."; // Shouldn't be used at all.

var phraseLen = finalPhrase.length; // Length of the target.

var chooseFrom = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789 !\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~";

// Function to generate a random string of length len.
function randomString(len) {
    var randomString = '';
    for (var p = 0; p < len; p++) {
    	var randomPoz = Math.floor(Math.random() * chooseFrom.length);
    	randomString += chooseFrom.substring(randomPoz,randomPoz+1);
    }
    return randomString;
}

// Function to populate the first generation with n creatures.
function populate(length) {
  var temp = [];
  for (var i = 0; i < length; i++) {
    var creature = randomString(phraseLen);
    temp[i] = creature;
  }
  return temp;
}

// Function to evaluate the fitness of a passed creature.
function evaluate(phrase, goal) {
  var fitness = 0;
  for (var a = 0; a < phrase.length; a++) {
    if (phrase[a] == goal[a]) {
      fitness++;
    }
  }
  fitness = fitness * fitness;
  return fitness;
}

// Function to mutate a word using a passed mutability multiplier.
function mutate(word, chance) {
  var wordList = [];
  var newWord = "";
  for (var a = 0; a < word.length; a++) {
    wordList.push(word[a]);
    // D&D-style chance calculation here. Lifted from Dan Shiffman's code.
    if (Math.random() < chance) {
      wordList[a] = randomString(1);
    }
    newWord += wordList[a]
  }
  return newWord;
}

// Function to reproduce given two parents.
// Uses the first half of the first parent and the second half of the second parent.
function reproduce(parent1, parent2) {
  var wordLen = parent1.length;
  if (wordLen > parent2.length) {
    return parent1;
  } else if (wordLen < parent2.length) {
    return parent2;
  }
  var offspring = "";
  for (var z = 0; z < wordLen; z++) {
    if (z < Math.ceil(wordLen/2)) {
      offspring += parent1[z];
    } else {
      offspring += parent2[z];
    }
  }
  offspring = mutate(offspring, mutability);
  return offspring;
}

// Function to delete the latter half of a passed list.
function killWorst(pop) {
  var killNum = Math.floor(pop.length/2);
  var survivors = pop;
  survivors.splice(pop.length - killNum, killNum);
  return survivors;
}

// Function to return a mating pool.
function makePool(vals, pop) {
  var matingPool = [];
  for (var a = 0; a < vals.length; a++) {
    for (var b = 0; b < vals[a]; b++) {
      matingPool.push(pop[b]);
    }
  }
  return matingPool;
}

// Function to return a random pair for mating.
function getPair(pool) {
  // This caused a problem for short target strings where none of the first generation
  // had a fitness greater than 0, so the matingPool was left empty.
  if (pool.length < 2) {
    pool.push(randomString(phraseLen));
    pool.push(randomString(phraseLen));
  }
  var mate1 = false;
  var mate2 = false;
  while ((!(mate1)) || (!(mate2))) {
    mate1 = pool[Math.floor(Math.random() * pool.length)];
    mate2 = pool[Math.floor(Math.random() * pool.length)];
  }
  pair = [];
  pair.push(mate1);
  pair.push(mate2);
  return pair;
}

// Function to update the elements on the page.
function draw(pop, vals) {
  document.getElementById("outputarea1").innerHTML = "Best: " + pop[0];
  document.getElementById("outputarea2").innerHTML = "Best fitness: " + vals[0];
  document.getElementById("outputarea3").innerHTML = "Generation: " + (genNum + 1);
}

// Function to iterate to the next generation given the current generation.
function nextGen(pop) {
  valuedPop = []; // TODO: make this persist though generations so we don't evaluate the same creature multiple times.
  for (var i = 0; i < pop.length; i++) {
    fitness = evaluate(pop[i], finalPhrase);
    valuedPop.push([fitness, pop[i]]);
  }
  delete i;
  valuedPop.sort(function(a, b){return b[0]-a[0]});
  delete pop;
  pop = []
  vals = []
  for (var i = 0; i < valuedPop.length; i++) {
    vals.push(valuedPop[i][0]);
    pop.push(valuedPop[i][1]);
  }
  delete i;
  delete valuedPop;
  setTimeout(draw(pop, vals), 100);
  //draw(pop, vals);
  console.log("Best: " + pop[0]);
  console.log("Best fitness: " + vals[0]);
  console.log("Generation: " + (genNum + 1));
  // If we are done.
  if (pop[0] == finalPhrase) {
    return [pop, false];
  }
  pop = killWorst(pop);
  vals = killWorst(vals);
  pop2 = [];
  // Only generate the pool once to reduce runtime.
  matingPool = makePool(vals, pop);
  for (var x = pop.length; x > 0; x--) {
    pair = getPair(matingPool);
    pop.push(reproduce(pair[0], pair[1]));
  }
  delete matingPool;
  for (var i = 0; i < pop.length; i++) {
    pop[i] = mutate(pop[i], mutability);
  }
  return [pop, true];
}

function nextIteration() {
  returned = nextGen(population);
  delete population;
  population = returned[0];
  cont = returned[1];
  delete returned;
  genNum++;
  if (cont) {
    setTimeout(nextIteration, 10);
  } else {
    alert("Completed!");
  }
}


// Do stuff here.

// Get the final phrase from the user.
var finalPhrase = prompt("Please enter a phrase\nto evolve towards.", "To be or not to be.");
var phraseLen = finalPhrase.length;
population = populate(popSize); // Make the first generation.
var cont = true;

setTimeout(nextIteration, 10);
//alert("Completed!");
